### ISSUE TITLE
Experimenting with Windows 10 and Cygwin

### ISSUE THREAD

**Issue by bmix** [pirate/bookmark-archiver#66]
Hello,
I managed to get as far as to have Chromium getting started on Windows 10 using Cygwin. But then it seems to hang. Here is what I have done:

1. Install *Cygwin* and open up a Bash.
2. Do *not* run `./setup.sh` since that assumes *apt* (which is a package manager on Debian based systems) or *Homebrew* (which is a macOS package manager). You must ensure all is installed manually.
3. Make sure Cygwin's Python3 is installed, along with all dependencies (I had to manually install the *request* module, You can do this with Cygwin's `setup.exe`).
4. Install Chromium (I also tried my pre-installed Chrome, but when it hung, I installed Chromium, since that is what @pirate supports officially). Use the *portable* install, to stay as close to an *ixish system as possible (no registry, etc.) May be unneeded, but...hey..! ;-)
5. In a Bash terminal type (your path may differ, of course, this is how I have it):
```console
$ ln -s /S/Applications/chrlauncher-win32-stable-codecs-sync/bin/chrome.exe /usr/local/bin/chromium-browser
```
`/usr/local/bin/` must be in Bash's path, of course

That should suffice to get you as far as I got:
```console
$ /usr/bin/python3 archive.py ./bookmark_files/test_pocket.html
[*] [2018-01-20 20:57:11] Adding 0 new links from ./bookmark_files/test_pocket.html to index
[i] [2018-01-20 20:57:11] Updating 3 links in archive index...
[*] Checking Dependencies:
/usr/local/bin/chromium-browser
```

Note, that this is the fourth run I had, so that may explain the *"Adding 0 new links..."* part. I created a very short `test_pocket.html` from the first few lines of my Pocket export.

I guess, it is right to assume, that @pirate does not use Windows and it's clear, that Windows is not supported, but I thought, it should be doable with Cygwin, maybe with a few modifications.

The issue at hand is, that *Chromium* (*Chrome* behaves the same) gets started, but in GUI mode. Then nothing more happens. It seems to hang somewhere. I checked, whether the Windows ports use the same command-line switches, and yes, it seems they do (I tested `chrome.exe --incognito` from DOS as well as from Bash (using the same symlink, `archive.py` uses), both successfully. I also tried adding my Chrome's user data dir to `CHROME_USER_DATA_DIR` as well as keeping it at `None`.

When I exit Chrome manually (the GUI is usable) `archive.py` will return with:
```console
[X] Failed to parse Chrome version, is it installed properly?
    Run ./setup.sh, then confirm it was installed with: chromium-browser --version
    See https://github.com/pirate/bookmark-archiver for help.
```

Did anybody else try this on Windows?

**Comment by pirate** [comment]
Try running this manually from cygwin:

```bash
chromium-browser --version
chromium-browser --headless --disable-gpu --screenshot http://example.com
```

**Comment by bmix** [comment]
```console
$ chromium-browser --version
```
The browser's GUI starts. No version information gets written to the console. I get the normal start-page. So, it's like a clean browser start from desktop.

```console
$ chromium-browser --headless --disable-gpu --screenshot http://heise.de
```
gives me following in `chrome_debug.log`:
```
[0121/131555.030:WARNING:render_frame_host_impl.cc(2646)] OnDidStopLoading was called twice.
[0121/131558.796:INFO:headless_shell.cc(568)] Written to file screenshot.png.
```
as well as a `screenshot.png` in Chromium's directory.

There is a launcher on Windows, which does a little setup, like defining `User Data` dir and some options. It seems to pass the commandline options to the main exec, when used with `archiver.py` the result is:

```console
$ /usr/bin/python3 archive.py ./bookmark_files/test_pocket.html
[*] [2018-01-21 15:08:37] Adding 0 new links from ./bookmark_files/test_pocket.html to index
[i] [2018-01-21 15:08:37] Updating 3 links in archive index...
[*] Checking Dependencies:
/usr/local/bin/chromium
[X] Failed to parse Chrome version, is it installed properly?
    Run ./setup.sh, then confirm it was installed with: chromium --version
    See https://github.com/pirate/bookmark-archiver for help.
```

So, for now, I do my testing with the `chrome.exe`, not with the launcher.

My config is:
```python
CHROME_BINARY =         os.getenv('CHROME_BINARY',         'chromium-browser' )
CHROME_USER_DATA_DIR =  os.getenv('CHROME_USER_DATA_DIR',  '/S/Applications/chrlauncher-win32-stable-codecs-sync/profile/Default')
```
To sum it up: When issuing a `--version` from the command-line, I get nothing in return. The browser starts its GUI and waits to be used. Nothing else.

When issuing `--headless --disable-gpu` and anything like `--screenshot` `--print-to-pdf` I get the desired result (headless start, action gets executed). However, doing ` --headless --disable-gpu --repl https://www.chromestatus.com/` does not seem to work. On DOS commandline the prompt returns to the DOS prompt, on Bash it does not return. In both cases I do not get the REPL prompt.

I am a bit in a hurry right now, just wanted to reply quickly, I will investigate deeper tonight.

**Comment by bmix** [comment]
(Not editing the last comment, instead writing a new one, not sure, whether Github notifies about comment-edits)

I forgot to mention the version of Chromium I am running (which I got from the *About* menu entry, not on the console):

> Version 63.0.3239.132 (Entwickler-Build) (32-Bit)

Here is a question by a user, who shared my experiences on Stack Overflow: [How do I use Headless Chrome in Chrome 60 on Windows 10?](https://stackoverflow.com/questions/45364102/how-do-i-use-headless-chrome-in-chrome-60-on-windows-10). They also talk about using Chrome programmatically with Python3 on Windows, I can do the step which installs the *WebDriver*, should you be interested.

Please note, that till now, I used your project from a Cygwin Bash with Cygwin's *python3*.

In order to step forward and get some more feel for what may be going on, I also tried it from the Windows DOS environment. For this I simply added the directory with `chrome.exe` to my path, to keep things as simple as possible (I have the encoding set that way, since I am on a German system with Umlatus and somehow the UTF-8 encoding in the console was no help).
```console
$ C:\Python35\python3.exe archive.py bookmark_files\test_pocket.html
[X] Your system is running python3 scripts with a bad locale setting: cp65001 (it should be UTF-8).
    To fix it, add the line "export PYTHONIOENCODING=utf8" to your ~/.bashrc file (without quotes)

    Confirm that it's fixed by opening a new shell and running:
        python3 -c "import sys; print(sys.stdout.encoding)"   # should output UTF-8

    Alternatively, run this script with:
        env PYTHONIOENCODING=utf8 ./archive.py export.html
[*] [2018-01-22 01:02:44] Adding 0 new links from bookmark_files\test_pocket.html to index
[i] [2018-01-22 01:02:44] Updating 3 links in archive index...
[*] Checking Dependencies:
/s/Applications/chrlauncher-win32-stable-codecs-sync/bin/chrome.exe

    --> Now Chrome has started in GUI mode ! After I quit it I get this: <--

[X] Failed to parse Chrome version, is it installed properly?
    Run ./setup.sh, then confirm it was installed with: chrome.exe --version
    See https://github.com/pirate/bookmark-archiver for help.
```

**Comment by pirate** [comment]
I feel like if you just remove the `chrome --version` check it might run fine.  Try commenting lines `util.py:59 - 74` and running it.

**Comment by bmix** [comment]
Yes, indeed, that did it!

Not sure, whether it's of interest to you, but here is what I got, when executing your script from a Windows DOS environment, using ActiveState's ActivePython3.5 (below that find the output under Cygwin):
```console
$ S:\Applications\bookmark-archiver>C:\Python35\python3.exe archive.py bookmark_files\test_pocket.html                               
[X] Your system is running python3 scripts with a bad locale setting: cp65001 (it should be UTF-8).                                
    To fix it, add the line "export PYTHONIOENCODING=utf8" to your ~/.bashrc file (without quotes)                                 
                                                                                                                                   
    Confirm that it's fixed by opening a new shell and running:                                                                    
        python3 -c "import sys; print(sys.stdout.encoding)"   # should output UTF-8                                                
                                                                                                                                   
    Alternatively, run this script with:                                                                                           
        env PYTHONIOENCODING=utf8 ./archive.py export.html                                                                         
[*] [2018-01-24 15:45:13] Adding 0 new links from bookmark_files\test_pocket.html to index                                         
[i] [2018-01-24 15:45:13] Updating 3 links in archive index...                                                                     
[*] Checking Dependencies:                                                                                                         
/s/Applications/chrlauncher-win32-stable-codecs-sync/bin/chrome.exe                                                                
/usr/bin/wget                                                                                                                      
/usr/bin/curl                                                                                                                      
[+] [1516329983] "Sh*tcoin vs. buttcoin: a dummies guide to cryptocurrency manipulation": counterthink.com/Shtcoin-Buttcoin.asp    
    html\archive\1516329983 (creating)                                                                                             
    - Fetching: wget                                                                                                               
[X] Index is up-to-date, archive update paused on link 0/2                                                                         
    Continue where you left off by running:                                                                                        
       ./archive.py bookmark_files\test_pocket.html 1516329983                                                                     
Traceback (most recent call last):                                                                                                 
  File "archive.py", line 124, in <module>                                                                                         
    update_archive(archive_path, links, source=source, resume=resume, append=True)                                                 
  File "archive.py", line 70, in update_archive                                                                                    
    archive_links(archive_path, links, source=source, resume=resume)                                                               
  File "S:\Applications\bookmark-archiver\archive_methods.py", line 60, in archive_links                                           
    raise e                                                                                                                        
  File "S:\Applications\bookmark-archiver\archive_methods.py", line 46, in archive_links                                           
    archive_link(link_dir, link)                                                                                                   
  File "S:\Applications\bookmark-archiver\archive_methods.py", line 79, in archive_link                                            
    link = fetch_wget(link_dir, link, overwrite=overwrite)                                                                         
  File "S:\Applications\bookmark-archiver\archive_methods.py", line 140, in timed_fetch_func                                       
    result = fetch_func(link_dir, link, **kwargs)                                                                                  
  File "S:\Applications\bookmark-archiver\archive_methods.py", line 183, in fetch_wget                                             
    end = progress(timeout, prefix='      ')                                                                                       
  File "S:\Applications\bookmark-archiver\util.py", line 157, in progress                                                          
    p.start()                                                                                                                      
  File "C:\Python35\lib\multiprocessing\process.py", line 105, in start                                                            
    self._popen = self._Popen(self)                                                                                                
  File "C:\Python35\lib\multiprocessing\context.py", line 212, in _Popen                                                           
    return _default_context.get_context().Process._Popen(process_obj)                                                              
  File "C:\Python35\lib\multiprocessing\context.py", line 313, in _Popen                                                           
    return Popen(process_obj)                                                                                                      
  File "C:\Python35\lib\multiprocessing\popen_spawn_win32.py", line 66, in __init__                                                
    reduction.dump(process_obj, to_child)                                                                                          
  File "C:\Python35\lib\multiprocessing\reduction.py", line 59, in dump                                                            
    ForkingPickler(file, protocol).dump(obj)                                                                                       
AttributeError: Can't pickle local object 'progress.<locals>.progress_bar'                                                         
                                                                                                                                   
S:\Applications\bookmark-archiver>Traceback (most recent call last):                                                               
  File "<string>", line 1, in <module>                                                                                             
  File "C:\Python35\lib\multiprocessing\spawn.py", line 100, in spawn_main                                                         
    new_handle = steal_handle(parent_pid, pipe_handle)                                                                             
  File "C:\Python35\lib\multiprocessing\reduction.py", line 81, in steal_handle                                                    
    _winapi.PROCESS_DUP_HANDLE, False, source_pid)                                                                                 
OSError: [WinError 87] Falscher Parameter                                                                                          
```
*Falscher Parameter* = *Wrong Parameter*

With the Cygwin version, most things were fine:
```console
$ /usr/bin/python3 ./archive.py bookmark_files/test_pocket.html
[*] [2018-01-24 15:47:12] Adding 0 new links from bookmark_files/test_pocket.html to index
[i] [2018-01-24 15:47:12] Updating 3 links in archive index...
[*] Checking Dependencies:
/s/Applications/chrlauncher-win32-stable-codecs-sync/bin/chrome.exe
/usr/bin/wget
/usr/bin/curl
[+] [1516329983] "Sh*tcoin vs. buttcoin: a dummies guide to cryptocurrency manipulation": counterthink.com/Shtcoin-Buttcoin.asp
    html/archive/1516329983 (creating)
    - Fetching: wget

    - Fetching: pdf
     ██████████                                                                                                                                                  - Fetching: screenshot


       Run to see full output: cd html/archive/1516329983; chrome.exe --headless --disable-gpu --user-data-dir=S:/Applications/chrlauncher-win32-stable-codecs-sync/profile/Default --screenshot --window-size=1440,900 http://counterthink.com/Shtcoin-Buttcoin.asp
       Failed: Exception Failed to chmod: screenshot.png does not exist (did the previous step fail?)
    - Fetching: archive_org

    - Fetching: favicon

 0.4% (0/60sec)    √ Updating: index.json
    √ Updating: index.html
[+] [1507559647] "https://www.brainpickings.org/2017/10/04/rollo-may-freedom-destiny-pause/": www.brainpickings.org/2017/10/04/rollo-may-freedom-destiny-pause/
    html/archive/1507559647 (creating)
    - Fetching: wget
███████
    - Fetching: pdf                                                                                                                                      4.4% (3/60sec)
          - Fetching: screenshot


       Run to see full output: cd html/archive/1507559647; chrome.exe --headless --disable-gpu --user-data-dir=S:/Applications/chrlauncher-win32-stable-codecs-sync/profile/Default --screenshot --window-size=1440,900 https://www.brainpickings.org/2017/10/04/rollo-may-freedom-destiny-pause/
       Failed: Exception Failed to chmod: screenshot.png does not exist (did the previous step fail?)
    - Fetching: archive_org
██
    - Fetching: favicon                                                                                                                                  1.6% (1/60sec)

 0.2% (0/60sec)    √ Updating: index.json
    √ Updating: index.html
[+] [1503082847] "TIL the Battle of the Eclipse, the last battle in a 6 year war between the ": www.reddit.com/r/todayilearned/comments/6ui3a7/til_the_battle_of_the_eclipse_the_last_battle_in/
    html/archive/1503082847 (creating)
    - Fetching: wget
    - Fetching: pdf

█████████████████                                                                                                                                           - Fetching: screenshot 11.4% (7/60sec)

       Run to see full output: cd html/archive/1503082847; chrome.exe --headless --disable-gpu --user-data-dir=S:/Applications/chrlauncher-win32-stable-codecs-sync/profile/Default --screenshot --window-size=1440,900 https://www.reddit.com/r/todayilearned/comments/6ui3a7/til_the_battle_of_the_eclipse_the_last_battle_in/?utm_source=ifttt
       Failed: Exception Failed to chmod: screenshot.png does not exist (did the previous step fail?)
    - Fetching: archive_org
      ██                                                                                                                                                       1.1% (1/60sec)    - Fetching: favicon

 0.3% (0/60sec)    √ Updating: index.json
    √ Updating: index.html
[√] [2018-01-24 15:48:12] Archive update complete (59.39 sec)
    - 0 entries skipped
    - 12 entries updated
    - 3 errors
```
